File: app/src/main/java/goodthingmap/android/prada/lab/goodthingmap/component/AlertDialogFragment.java
package goodthingmap.android.prada.lab.goodthingmap.component;

/**
 * Created by prada on 2014/7/5.
 */

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.text.TextUtils;

import com.afollestad.materialdialogs.AlertDialogWrapper;
import com.afollestad.materialdialogs.AlertDialogWrapper.Builder;

public class AlertDialogFragment extends DialogFragment {

    private static final String KEY_TITLE = "title";
    private static final String KEY_MESSAGE = "message";
    private static final String KEY_POSITIVE_BUTTON_TITLE = "positive_button_title";
    private static final String KEY_NEGATIVE_BUTTON_TITLE = "negative_button_title";

    private DialogInterface.OnClickListener mPositiveListener;
    private DialogInterface.OnClickListener mNegativeListener;

    /**
     * Creates a new instance of AlertDialogFragment.
     * @param title string. The dialog won't have title if it's empty.
     * @param message string. The dialog won't have message if it's empty.
     * @param positiveButtonTitle Positive button title string. Required.
     * @param positiveListener Listener for the positive button. Required.
     * @return
     */
    public static AlertDialogFragment newInstance(String title,
                                                  String message,
                                                  String positiveButtonTitle,
                                                  DialogInterface.OnClickListener positiveListener) {
        return newInstance(title, message, positiveButtonTitle, positiveListener,
                null, null);
    }

    /**
     * Creates a new instance of AlertDialogFragment.
     * @param title string. The dialog won't have title if it's empty.
     * @param message string. The dialog won't have message if it's empty.
     * @return
     */
    public static AlertDialogFragment newInstance(String title,
                                                  String message,
                                                  String positiveButtonTitle,
                                                  DialogInterface.OnClickListener positiveListener,
                                                  String negativeButtonTitle,
                                                  DialogInterface.OnClickListener negativeListener) {
        AlertDialogFragment fragment = new AlertDialogFragment();

        Bundle bundle = new Bundle();
        bundle.putString(KEY_TITLE, title);
        bundle.putString(KEY_MESSAGE, message);
        bundle.putString(KEY_POSITIVE_BUTTON_TITLE, positiveButtonTitle);
        bundle.putString(KEY_NEGATIVE_BUTTON_TITLE, negativeButtonTitle);
        fragment.setArguments(bundle);

        fragment.setPositiveListener(positiveListener);
        fragment.setNegativeListener(negativeListener);

        return fragment;
    }

    public void setPositiveListener(DialogInterface.OnClickListener mListener) {
        this.mPositiveListener = mListener;
    }

    public void setNegativeListener(DialogInterface.OnClickListener mListener) {
        this.mNegativeListener = mListener;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        String title = args.getString(KEY_TITLE);
        String message = args.getString(KEY_MESSAGE);
        String positiveButtonTitle = args.getString(KEY_POSITIVE_BUTTON_TITLE);

        Builder builder = new AlertDialogWrapper.Builder(getActivity())
                .setPositiveButton(positiveButtonTitle, mPositiveListener)
                .setCancelable(false);

        String negativeButtonTitle = args.getString(KEY_NEGATIVE_BUTTON_TITLE);
        if (!TextUtils.isEmpty(negativeButtonTitle)) {
            builder.setNegativeButton(negativeButtonTitle, mNegativeListener);
        }
        if (!TextUtils.isEmpty(title)) {
            builder.setTitle(title);
        }
        if (!TextUtils.isEmpty(message)) {
            builder.setMessage(message);
        }
        return builder.create();
    }
}

File: app/src/main/kotlin/goodthingmap/android/prada/lab/goodthingmap/component/AlertDialogFragment.kt
package goodthingmap.android.prada.lab.goodthingmap.component

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.TextUtils
import androidx.fragment.app.DialogFragment
import com.afollestad.materialdialogs.AlertDialogWrapper

class AlertDialogFragment : DialogFragment() {
    companion object {
        private const val KEY_TITLE = "title"
        private const val KEY_MESSAGE = "message"
        private const val KEY_POSITIVE_BUTTON_TITLE = "positive_button_title"
        private const val KEY_NEGATIVE_BUTTON_TITLE = "negative_button_title"

        @JvmStatic
        fun newInstance(
            title: String?,
            message: String?,
            positiveButtonTitle: String?,
            positiveListener: DialogInterface.OnClickListener
        ): AlertDialogFragment = newInstance(
            title, message, positiveButtonTitle, positiveListener, null, null
        )

        @JvmStatic
        fun newInstance(
            title: String?,
            message: String?,
            positiveButtonTitle: String?,
            positiveListener: DialogInterface.OnClickListener?,
            negativeButtonTitle: String?,
            negativeListener: DialogInterface.OnClickListener?
        ): AlertDialogFragment = AlertDialogFragment().apply {
            arguments = Bundle().apply {
                putString(KEY_TITLE, title)
                putString(KEY_MESSAGE, message)
                putString(KEY_POSITIVE_BUTTON_TITLE, positiveButtonTitle)
                putString(KEY_NEGATIVE_BUTTON_TITLE, negativeButtonTitle)
            }
            this.positiveListener = positiveListener
            this.negativeListener = negativeListener
        }
    }

    private var positiveListener: DialogInterface.OnClickListener? = null
    private var negativeListener: DialogInterface.OnClickListener? = null

    fun setPositiveListener(listener: DialogInterface.OnClickListener?) {
        positiveListener = listener
    }

    fun setNegativeListener(listener: DialogInterface.OnClickListener?) {
        negativeListener = listener
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = requireArguments()
        val builder = AlertDialogWrapper.Builder(requireActivity())
            .setPositiveButton(args.getString(KEY_POSITIVE_BUTTON_TITLE).orEmpty(), positiveListener)
            .setCancelable(false)
        args.getString(KEY_NEGATIVE_BUTTON_TITLE)?.takeUnless(TextUtils::isEmpty)?.let {
            builder.setNegativeButton(it, negativeListener)
        }
        args.getString(KEY_TITLE)?.takeUnless(TextUtils::isEmpty)?.let(builder::setTitle)
        args.getString(KEY_MESSAGE)?.takeUnless(TextUtils::isEmpty)?.let(builder::setMessage)
        return builder.create()
    }
}
